frostsnap_core/tests/proptest.rs

```diff
@@ -0,0 +1,723 @@
+                        .do_keygen
+                        .device_to_share_index
+                        .contains_key(device_id)
+                        && !keygen_state.devices_confirmed.contains(device_id)
+                }
+                None => false,
+            },
+            Transition::CKeygenConfirm { keygen_id } => {
+                match state.pending_keygens.get(keygen_id) {
+                    Some(keygen_state) => {
+                        keygen_state.devices_confirmed.len()
+                            == keygen_state.do_keygen.device_to_share_index.len()
+                    }
+                    None => false,
+                }
+            }
+            Transition::CNonceReplenish { device_id } => {
+                state.run_start.device_set().contains(device_id)
+            }
+            Transition::CStartSign {
+                key_index, devices, ..
+            } => match state.finished_keygens.get(*key_index) {
+                Some(keygen) => {
+                    !keygen.deleted
+                        && keygen.do_keygen.devices().is_superset(devices)
+                        && state.available_signing_devices().is_superset(devices)
+                }
+                None => false,
+            },
+            Transition::CSendSignRequest {
+                session_index,
+                device_id,
+            } => match state.sign_sessions.get(*session_index) {
+                Some(session) => {
+                    session.devices.contains(device_id)
+                        && session.key_index < state.finished_keygens.len()
+                        && !state.finished_keygens[session.key_index].deleted
+                        && !session.canceled
+                }
+                None => false,
+            },
+            Transition::CCancelSignSession { session_index } => {
+                match state.sign_sessions.get(*session_index) {
+                    Some(session) => {
+                        session.key_index < state.finished_keygens.len()
+                            && !state.finished_keygens[session.key_index].deleted
+                            && !session.canceled
+                    }
+                    None => false,
+                }
+            }
+            Transition::DAckSignRequest {
+                session_index,
+                device_id,
+            } => match state.sign_sessions.get(*session_index) {
+                Some(session) => {
+                    session.sent_req_to.contains(device_id)
+                        && session.key_index < state.finished_keygens.len()
+                        && !state.finished_keygens[session.key_index].deleted
+                        && !session.got_sigs_from.contains(device_id)
+                        && !session.canceled
+                }
+                None => false,
+            },
+            &Transition::CDeleteKey { key_index } => key_index < state.finished_keygens.len(),
```

Comment:
Could also check `!state.finished_keygens[key_index].deleted`, but I tested and it doesn't bring any failures